README.md

```diff
@@ -1,20 +1,105 @@
-This is a prototype for a password delivery system from a service provider to any users that need secrets.
+This is a text delivery system that someone can utilize to securely send shared
+secrets or password or even just arbitrary text to someone else.
 
-This system has the downside of having credentials at rest in a recoverable way, in the database.
+The system tries to minimize the attack surface on the secrets by encrypting
+data at rest with a token that isn't stored on the system, and the token is only
+provided to the system by a user requesting the data associated with it, and
+when retrieving the data it is deleted as part of the same database query.
 
-## TODO
+## Goals
 
-* Password "masking" in the database
-  * Hashing is not possible due to needing plaintext
-  * Encrypting is possible but the server needs the plaintext so maybe it's useless
-  * Maybe a simple way is just using base64?
-* Actual CSS
-* If the frontend is exposed to public, some rate limit methods should be investigated
+Maintain project simplicity while also securely handling secrets with reversible
+encryption.
 
-## Goals
+## How?
+
+When an entry is created, it gets a Version 4 UUID generated as a unique key
+to find it later.
+
+The UUID is then used as a key for PostgreSQLs `pgp_sym_encrypt`* function to
+encrypt the actual secret data.
+
+But we don't want to store the decryption key right next to the data, so to
+mask it from the database, we pass the UUID, along with a site wide secret
+(salt?), to PostgreSQLs `hmac`* function and store the result as our lookup key.
+
+By doing things this way, we don't have the decryption key for any entry unless
+a request from a user provides it to us, we then pass it to `hmac` again to find
+an entry that matches the same key hash, and finally using the UUID as a key, we
+decrypt the payload field and return it to the requesting client.
+
+## Security
+
+The best case scenario to transmit a secret to someone is to have already
+established a public/private key pair with them, then it is trivial to just
+encrypt and sign the data and send it over any medium, but in reality, very few
+people have those or even the knowledge how to use them (securely).
+
+The attack surface on the shared secrets in this system should only be around
+submission and retrieval of the secrets since we have to work with them in
+plain text.
+
+Our biggest vulnerability is unnoticed front end replacement/hijack while the
+system is in active use.
+
+Database dump/theft is a low priority problem since you need the UUID to decrypt
+the payload field.
+
+A plausible attack scenario is if an attacker can get a database backup _and_
+the request/ticket between the provider and user to get the UUID and what system
+that secret is for.
+
+But then you'd have bigger issues to think about.
+
+### Database
+
+The front end application accesses the database with a very restricted user, it
+only has `execute` permission on our two functions, nothing else.
```

Comment:
`s/on our two functions/on two stored procedures/`

Reply:
Fixed in next push